### ISSUE TITLE
Literal[False] + Literal[True] does not full cover bool type?

### ISSUE THREAD

**Issue by Holt59** [microsoft/pyright#8135]
**Describe the bug**

When a function has two overloads, one taking `Literal[True]` and the other `Literal[False]`, pyright throws an error when trying to call it with a `bool` value or even with a `Literal[True, False]` value.

Changing `bool` in the implementation of `f1` below to `Literal[True, False]` does not change the behavior. Using non-bool literals does not produce the issue.

**Code or Screenshots**

```python
from typing import Literal, overload

@overload
def f1(c: Literal[False] = ...) -> str: ...

@overload
def f1(c: Literal[True] = ...) -> int: ...

def f1(c: bool = False) -> str | int: ...

def r_bool() -> bool: ...
def r_literal() -> Literal[True, False]: ...

f1(r_bool())  # no overload matches for f1
f1(r_literal())  # no overload matches for f1

```

**VS Code extension or command-line**

Both, pyright CLI is 1.1.367

I do not remember seeing that problem before so maybe it is a recent update.

**Comment by erictraut** [comment]
This is a duplicate of https://github.com/microsoft/pyright/issues/5230.

**Comment by Holt59** [comment]
This still raises an error and there is no use of `bool`

```python
from typing import Literal, overload

def f1(c: Literal[False] = ...) -> str: ...
@overload
def f1(c: Literal[True] = ...) -> int: ...
def f1(c: Literal[True, False] = ...) -> str | int: ...

def r_literal() -> Literal[True, False]: ...

f1(r_literal())
```

In the linked issue you said

> Pyright doesn't expand bool into Literal[True, False] in this case. That's an intentional decision, as expansion of this sort would be computationally expensive in a hot path within the type analyzer.

Which I understand, but if you do not expand `bool` into `Literal[True, False]`, why collapse `Literal[True, False]` into `bool` (in the above exxample `r_literal()` is deduced to return `bool`)?
